### DISCUSSION TITLE
Disappearing implicit hydrogen in TautomerEnumerator canonicalization

### DISCUSSION THREAD

**Discussion by gedeck** [rdkit/rdkit#5431] [Q&A]
I made the following observation when using the TautomerEnumerator. 

Boilerplate:
```
from rdkit import Chem
from rdkit.Chem.MolStandardize import rdMolStandardize
enumerator = rdMolStandardize.TautomerEnumerator()
```
achiral molecule as reference  (Alanine)
```
mol = Chem.MolFromSmiles('CC(N)C(O)=O')
mol = enumerator.Canonicalize(mol)
mol.Debug()
```
gives
```
Atoms:
	0 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
	1 6 C chg: 0  deg: 3 exp: 3 imp: 1 hyb: 4 arom?: 0 chi: 0
	2 7 N chg: 0  deg: 1 exp: 1 imp: 2 hyb: 4 arom?: 0 chi: 0
	3 6 C chg: 0  deg: 3 exp: 4 imp: 0 hyb: 3 arom?: 0 chi: 0
	4 8 O chg: 0  deg: 1 exp: 1 imp: 1 hyb: 3 arom?: 0 chi: 0
	5 8 O chg: 0  deg: 1 exp: 2 imp: 0 hyb: 3 arom?: 0 chi: 0
```
Focus on the implicit hydrogens: `(3, 1, 2, 0, 1, 0)`

Now using a chiral molecule
```
mol = Chem.MolFromSmiles('C[C@H](N)C(O)=O')
mol.Debug()
```
The implicit hydrogens are as expected: `(3, 0, 2, 0, 1, 0)`
```
Atoms:
	0 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
	1 6 C chg: 0  deg: 3 exp: 4 imp: 0 hyb: 4 arom?: 0 chi: 2
	2 7 N chg: 0  deg: 1 exp: 1 imp: 2 hyb: 4 arom?: 0 chi: 0
	3 6 C chg: 0  deg: 3 exp: 4 imp: 0 hyb: 3 arom?: 0 chi: 0
	4 8 O chg: 0  deg: 1 exp: 1 imp: 1 hyb: 3 arom?: 0 chi: 0
	5 8 O chg: 0  deg: 1 exp: 2 imp: 0 hyb: 3 arom?: 0 chi: 0
```
Now generate the canonical tautomer:
```
mol = enumerator.Canonicalize(mol)
mol.Debug()
```
The implicit hydrogen on the OH of the carboxylic acid is now set to 0 `(3, 0, 2, 0, 0, 0)`
```
Atoms:
	0 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
	1 6 C chg: 0  deg: 3 exp: 4 imp: 0 hyb: 4 arom?: 0 chi: 0
	2 7 N chg: 0  deg: 1 exp: 1 imp: 2 hyb: 4 arom?: 0 chi: 0
	3 6 C chg: 0  deg: 3 exp: 4 imp: 0 hyb: 3 arom?: 0 chi: 0
	4 8 O chg: 0  deg: 1 exp: 2 imp: 0 hyb: 3 arom?: 0 chi: 0
	5 8 O chg: 0  deg: 1 exp: 2 imp: 0 hyb: 3 arom?: 0 chi: 0
```
The SMILES for mol is fine: `CC(N)C(=O)O`, but the implicit hydrogens don't correspond to this SMILES.

I don't see this problem with beta-alanine, so I guess the issue is with the nitrogen next to the alpha carbon. 

Is it a bug or am I missing a flag in the TautomerEnumerator?

**Comment by greglandrum** [comment]
@gedeck the confusion here is due to the RDKit's unfortunate distinction between two types of hydrogens which are not present in the molecular graph (what you would rationally think of as "implicit" hydrogens). In the RDKit's API there are "explicit" and "implicit" Hs not present in the graph. The first contribute to the explicit valence displayed by the debug command, the second to the implicit valence. The "explicit" Hs arise from Hs which are included in square brackets in SMILES

You can see the difference here:
```
In [11]: Chem.MolFromSmiles("OC[OH]").Debug()
Atoms:
	0 8 O chg: 0  deg: 1 exp: 1 imp: 1 hyb: 4 arom?: 0 chi: 0
	1 6 C chg: 0  deg: 2 exp: 2 imp: 2 hyb: 4 arom?: 0 chi: 0
	2 8 O chg: 0  deg: 1 exp: 2 imp: 0 hyb: 4 arom?: 0 chi: 0
Bonds:
	0 0->1 order: 1 conj?: 0 aromatic?: 0
	1 1->2 order: 1 conj?: 0 aromatic?: 0
```
atom 0 has an explicit valence of 1 and an implicit valence of 1, while atom 2 has an explicit valence of 2 and an implicit valence of 0.

If you want to find out the number of Hs on an atom, I would suggest using the function `GetTotalNumHs()`, which will return the sum of the "explicit" and "implicit" H counts. If you provide the optional argument `includeNeighbors=True` then it will also look for Hs which are actually in the graph.

To be very clear: I think this distinction was a mistake and if I could start over again from scratch this distinction would *not* be there. I have periodically toyed with the idea of fixing the situation, but doing so could end up breaking a lot of other people's code and I'm very reluctant to do that.
